### PR TITLE
revise: clarifies language around the allowed number of sections in tasks and workflows

### LINKED ISSUE #594: Clarify which task elements and workflows are singular and which can be specified multiple times.
There a quite few workflow and task elements that the specification _appears_ to imply can only occur once but that the grammar allows multiple occurrences.

For example, take a look at the task elements:

- An [input section](https://github.com/openwdl/wdl/blob/wdl-1.1/SPEC.md#task-inputs) appears to be referred to as a singular section.
- A [private declaration](https://github.com/openwdl/wdl/blob/wdl-1.1/SPEC.md#private-declarations) appears to allow zero or more declarations.
- A [command section](https://github.com/openwdl/wdl/blob/wdl-1.1/SPEC.md#command-section) appears to be referred to as a singular section.
- An [output section](https://github.com/openwdl/wdl/blob/wdl-1.1/SPEC.md#task-outputs) appears to be referred to as a singular section.
- A [runtime section](https://github.com/openwdl/wdl/blob/wdl-1.1/SPEC.md#runtime-section) appears to be referred to as a singular section.
- Both [metadata and parameter metadata sections](https://github.com/openwdl/wdl/blob/wdl-1.1/SPEC.md#metadata-sections) appear to be referred to a singular, respectively.

For what it's worth, every example I have seen appears to follow the above conventions.

Contrast that with the [parser implementation](https://github.com/openwdl/wdl-parsers/blob/main/antlr4/1.1/WdlV1_1Parser.g4#L247C1-L247C1), which I interpret as (paraphrasing): a task can contain one or more task elements. A task element is defined as a `task_input` section, a `task_output` section, a `task_command` section, a `task_runtime` section, a `bound_decls` section, a `parameter_meta`section, or a `meta` section. As such, it follows that these elements can be defined multiple times.

### PR DESCRIPTION
**Pull request by claymcleod**
This pull request clarifies the allowed number of member sections within `task` and `workflow` blocks.

Rendered [task](https://github.com/claymcleod/wdl/blob/spec/clarify-number-of-sections/SPEC.md#task-definition) and [workflow](https://github.com/claymcleod/wdl/blob/spec/clarify-number-of-sections/SPEC.md#workflow-elements) sections.

It wasn't clear to me whether I should update bump the patch version number myself or whether the maintainers of the spec would bundle this change in with others—let me know if you'd like me to do that.

Closes #594.

### Checklist
- [x] Pull request details were added to CHANGELOG.md

### CODE REVIEW

File: SPEC.md [hunk 1]
```diff
@@ -113,8 +113,8 @@ Revisions to this specification are made periodically in order to correct errors
     - [Fully Qualified Names \& Namespaced Identifiers](#fully-qualified-names--namespaced-identifiers)
     - [Call Statement](#call-statement)
       - [Computing Call Inputs](#computing-call-inputs)
-    - [Scatter](#scatter)
-    - [Conditional (`if`)](#conditional-if)
+    - [Scatter Statement](#scatter-statement)
```
Comment (claymcleod) [inline SPEC.md:128]: These sections were normalized to be consistent with "Call Statement".

File: SPEC.md [hunk 2]
```diff
@@ -2997,13 +2997,19 @@ A WDL task can be thought of as a template for running a set of commands - speci
 
 A task is defined using the `task` keyword, followed by a task name that is unique within its WDL document.
 
-A task has a required [`command`](#command-section) that is a template for a Bash script.
+Tasks are comprised of the following elements:
 
-Tasks explicitly define their [`input`s](#task-inputs) and [`output`s](#task-outputs), which is essential for building dependencies between tasks and workflows. The value of an input declaration may be supplied by the caller. Tasks may have additional "private" declarations within the task body. All task declarations may be initialized with hard-coded literal values, or may have their values constructed from expressions. Input and private declarations can be referenced in the command template.
+* A singular, required [`command`](#command-section), which defines the Bash script to be executed.
```
Comment (claymcleod) [inline SPEC.md]: A concise, bulleted list is easier for would-be implementers to parse. Many of the descriptions within this section were duplicated content from within the respective sections, so I cleaned that up with simple summaries that point to the respective sections for more information.

File: SPEC.md [hunk 3]
```diff
@@ -2997,13 +2997,19 @@ A WDL task can be thought of as a template for running a set of commands - speci
...
+* A singular, optional [`input`](#task-inputs) section that defines the inputs for the task.
+* A singular, optional [`output`](#task-outputs) section that defines the outputs for the task.
+* A singular, optional [`runtime`](#runtime-section) section that defines the required runtime environment conditions required for the task to execute properly.
+* A singular, optional [`meta`](#metadata-sections) section that defines task-level metadata.
+* A singular, optional [`parameter_meta`](#parameter-metadata-section) section that defines parameter-level metadata.
+* Any number of [private declarations](#private-declarations), which are declarations within the task body.
 
-A task may also specify [requirements for the runtime environment](#runtime-section) (such as the amount of RAM or number of CPU cores) that must be satisfied in order for its commands to execute properly.
+There is no enforced ordering of task elements.
```
Comment (claymcleod) [inline SPEC.md]: Explicitly state that there is no ordering of elements within tasks.

File: SPEC.md [hunk 4]
```diff
@@ -5019,7 +5023,7 @@ Test config:
 
 ## Workflow Definition
 
-A workflow can be thought of as a directed acyclic graph (DAG) of transformations that convert the input data to the desired outputs. Rather than explicitly specifying the sequence of operations, A WDL workflow instead describes the connections between the steps in the workflow (i.e., between the nodes in the graph). It is the responsibility of the execution engine to determine the proper ordering of the workflow steps, and to orchestrate the execution of the different steps.
+A workflow can be thought of as a directed acyclic graph (DAG) of transformations that convert the input data to the desired outputs. Rather than explicitly specifying the sequence of operations, a WDL workflow instead describes the connections between the steps in the workflow (i.e., between the nodes in the graph). It is the responsibility of the execution engine to determine the proper ordering of the workflow steps, and to orchestrate the execution of the different steps.
```
Comment (claymcleod) [inline SPEC.md:5485]: Very small, unrelated fix here.

File: SPEC.md [hunk 5]
```diff
@@ -5053,19 +5057,21 @@ workflow name {
 
 ### Workflow Elements
 
-Tasks and workflows have several elements in common. These sections have nearly the same usage in workflows as they do in tasks, so we just link to their earlier descriptions.
+Tasks and workflows have several elements in common. When applicable, the task definition for these sections is linked to rather than duplicated.
```
Comment (claymcleod) [inline SPEC.md:5523]: I felt the wording for this section was a bit too verbose. This wording simplifies things IMO:

- Here is a list of sections, as with tasks.
- We state much more simply that workflows have common elements with tasks and that we will link back to them.

File: SPEC.md [hunk 6]
```diff
@@ -5053,19 +5057,21 @@ workflow name {
...
-* [`meta` section](#metadata-sections)
-* [`parameter_meta` section](#parameter-metadata-section)
+A workflow is comprised of the following elements:
 
-In addition to these sections, a workflow may have any of the following elements that are specific to workflows:
+* A singular, optional [`input`](#task-inputs) section (_identfical to the `input` section within tasks_).
+* A singular, optional [`output`](#task-outputs) section (_identfical to the `output` section within tasks_).
+* A singular, optional [`meta`](#metadata-sections) section (_identfical to the `meta` section within tasks_).
+* A singular, optional [`parameter_meta`](#parameter-metadata-section) section (_identfical to the `parameter_meta` section within tasks_).
+* Any number of workflow execution elements, which include the following:
```
Comment (claymcleod) [inline SPEC.md:5528]: This introduces the term "workflow execution elements", which I would propose is what we call these execution-based elements. I considered using the term "workflow execution statements", but I just stuck with "elements" to be consistent with the language I used for tasks.

File: SPEC.md [hunk 7]
```diff
@@ -5053,19 +5057,21 @@ workflow name {
...
+  * A [private declaration](#private-declarations) (_identical to the private declarations within tasks_).
+  * A [`call`](#call-statement) statement, which invokes tasks or subworkflows.
+  * A [`scatter`](#scatter-statement) statement, which enables parallelized of workflow execution elements across collections.
+  * A [conditional (`if`)](#conditional-statement) statement, which enables conditional execution of workflow execution elements.
 
-* [`call`s](#call-statement) to tasks or subworkflows
-* [`scatters`](#scatter), which are used to parallelize operations across collections
-* [Conditional (`if`)](#conditional-if-block) statements, which are only executed when a conditional expression evaluates to `true`
+There is no enforced ordering of workflow elements.
```
Comment (claymcleod) [inline SPEC.md]: Explicitly state that there is no ordering of elements within workflows.

**Comment by patmagee** [comment]
@jdidion For clarifications like this, should we make them against 1.1.2? or against 1.1?

**Comment by jdidion** [comment]
@claymcleod this is a great PR, thanks!

The only question is where to merge. Historically, our policy has been not to update a spec once it has been released. We recently modified this to allow patch releases, and have just released `1.1.1`.

If we want to do another patch release to `1.1` (i.e., `1.1.2`), then we should create a new branch for that (`wdl-1.1.2`), and this PR should target that branch. Otherwise, this PR should target `wdl-1.2`.

If it is decided that this branch should target `wdl-1.2`, then it will need to be updated to recognize the new `requirements` and `hints` section, as well as the deprecation of `runtime`.

In my opinion, we should get in the habit of making frequent patch releases (perhaps on a quarterly cadence, assuming there are any patches to make). So I suggest option 1 (targeting a new `wdl-1.1.2` branch). Keep in mind that whenever a patch release is made, it is merged into the minor release branch, and then the next minor release branch will need to be rebased. In this case: merge `wdl-1.1.2` into `wdl-1.1`, then rebase `wdl-1.2` to `wdl-1.1@HEAD`.

Would appreciate feedback on this from others before going forward.

File: SPEC.md (hunk 7 above)
Comment (geoffjentry) [reply SPEC.md]: I think it'd be good to include a bit suggesting people order them in a chronological fashion to improve readability.

**Comment by geoffjentry** [comment]
+1 to `1.1.2`

**Comment by claymcleod** [comment]
Yeah, I'm also in favor of `1.1.2`.

File: SPEC.md (hunk 7 above)
Comment (claymcleod) [reply SPEC.md]: I have updated the description in [73b0033](https://github.com/openwdl/wdl/pull/598/commits/73b0033be1c8cbc8812e64c94e5a4584e311f007), let me know what you think.

**Comment by jdidion** [comment]
Moved to `wdl-1.1.2`

**Review by jdidion — APPROVED** [review]

File: SPEC.md [hunk 8]
```diff
@@ -2997,13 +2997,17 @@ A WDL task can be thought of as a template for running a set of commands - speci
...
 A task is defined using the `task` keyword, followed by a task name that is unique within its WDL document.
 
-A task has a required [`command`](#command-section) that is a template for a Bash script.
+Tasks are comprised of the following elements:
 
-Tasks explicitly define their [`input`s](#task-inputs) and [`output`s](#task-outputs), which is essential for building dependencies between tasks and workflows. The value of an input declaration may be supplied by the caller. Tasks may have additional "private" declarations within the task body. All task declarations may be initialized with hard-coded literal values, or may have their values constructed from expressions. Input and private declarations can be referenced in the command template.
+* A singular, optional [`input`](#task-inputs) section, which defines the inputs for the task.
+* A singular, required [`command`](#command-section), which defines the Bash script to be executed.
+* A singular, optional [`output`](#task-outputs) section, which defines the outputs for the task.
+* A singular, optional [`runtime`](#runtime-section) section, which defines the required runtime environment conditions.
```
Comment (jdidion) [inline SPEC.md]: We are deprecating `runtime` in favor of `requirements` and `hints` but that PR (#595) isn't merged yet. I think it makes sense to merge this PR first and then updated the above in #595.

File: SPEC.md [hunk 9]
```diff
@@ -2997,13 +2997,17 @@ A WDL task can be thought of as a template for running a set of commands - speci
...
+* A singular, optional [`meta`](#metadata-sections) section, which defines task-level metadata.
+* A singular, optional [`parameter_meta`](#parameter-metadata-section) section, which defines parameter-level metadata.
+* Any number of [private declarations](#private-declarations).
 
-A task may also specify [requirements for the runtime environment](#runtime-section) (such as the amount of RAM or number of CPU cores) that must be satisfied in order for its commands to execute properly.
-
-There are two optional metadata sections: the [`meta`](#metadata-sections) section, for task-level metadata, and the [`parameter_meta`](#parameter-metadata-section) section, for parameter-level metadata.
+Although there is no enforced order for task elements, it is recommended to organize them logically, starting with execution-based elements in chronological order, then specifying runtime requirements, and finally, adding metadata about the task.
```
Comment (jdidion) [inline SPEC.md]: This seems more suited to a style guide. I don't think it belongs in the spec.

I might support more neutral language such as "...it is recommended to follow a consistent pattern for all the tasks in a workflow."

File: SPEC.md [hunk 10]
```diff
@@ -5053,19 +5057,21 @@ workflow name {
...
-In addition to these sections, a workflow may have any of the following elements that are specific to workflows:
+* A singular, optional [`input`](#task-inputs) section (_identical to the `input` section within tasks_).
+* Any number of workflow execution elements, which include the following:
+  * A [private declaration](#private-declarations) (_identical to private declarations within tasks_).
+  * A [`call`](#call-statement) statement, which invokes tasks or subworkflows.
+  * A [`scatter`](#scatter-statement) statement, which enables parallelized of workflow execution elements across collections.
+  * A [conditional (`if`)](#conditional-statement) statement, which enables conditional execution of workflow execution elements.
+* A singular, optional [`output`](#task-outputs) section (_identical to the `output` section within tasks_).
+* A singular, optional [`meta`](#metadata-sections) section (_identical to the `meta` section within tasks_).
+* A singular, optional [`parameter_meta`](#parameter-metadata-section) section (_identical to the `parameter_meta` section within tasks_).
```
Comment (jdidion) [inline SPEC.md:5536]: Also need to add `hints` here, but that can be done in #595.

File: SPEC.md [hunk 11]
```diff
@@ -5053,19 +5057,21 @@ workflow name {
...
 
-* [`call`s](#call-statement) to tasks or subworkflows
-* [`scatters`](#scatter), which are used to parallelize operations across collections
-* [Conditional (`if`)](#conditional-if-block) statements, which are only executed when a conditional expression evaluates to `true`
+Although there is no enforced order for workflow elements, it is recommended to organize them logically, starting with execution-based elements in chronological order followed by metadata about the workflow.
```
Comment (jdidion) [inline SPEC.md]: Same comment here as above.

**Review by jdidion — CHANGES_REQUESTED** [review]
Thanks @claymcleod. Overall I think this is great; however, I don't think we should be adding style recommendations here. Note that I haven't gone through the spec thoroughly to make sure I'm being consistent with this sentiment, but if there are places in the spec already where we're making stylistic recommendations I'd also advocate for moving those to a separate style guide.

File: SPEC.md (hunk 9 above)
Comment (jdidion) [reply SPEC.md]: I now realize this conflicts with @geoffjentry comment above. I'll bring this up on Slack and try to get consensus.

Comment (geoffjentry) [reply SPEC.md]: @jdidion Agree a style guide would be more appropriate.

File: SPEC.md (hunk 8 above)
Comment (claymcleod) [reply SPEC.md]: I have updated this and added a 🗑️ to the `runtime` section so it can be removed at the same time as the section itself.

File: SPEC.md (hunk 9 above)
Comment (claymcleod) [reply SPEC.md]: I have addressed these by simply stating there is no order to the elements.

**Comment by claymcleod** [comment]
@jdidion all comments have been addressed, the branch target has been updated to `wdl-1.2`, and, I believe at least, the changes necessary to make this merge-able into `wdl-1.2` have been made. Please take a look and let me know if we need anything else.

**Comment by jdidion** [comment]
@claymcleod Regarding the target branch, maybe there was a misunderstanding. I think we agreed to target `wdl-1.1.2`. Can you please change it back?

Otherwise it looks good and I'm prepared to merge it.

**Comment by jdidion** [comment]
Made these changes here https://github.com/openwdl/wdl/pull/632 rather than trying to rebase

**Comment by claymcleod** [comment]
Hey John,

Apologies, I was planning on responding last week but then was out.

In short, I had interpreted this comment:

> If it is decided that this branch should target wdl-1.2, then it will need to be updated to recognize the new requirements and hints section, as well as the deprecation of runtime.

To mean that it would be okay to merge into 1.2 if I was willing to do the work to update the sections to be compliant with 1.2. Originally, I had agreed with that, but then I found some time to just update it for 1.2.

I'm okay with it going into either 1.1.2 or 1.2 provided that we make sure the right content gets on the right branch. In this case, the text you had merged in contains references to the `requirements` and `hints` sections while referring to the `runtime` section as 🗑️. Ultimately, the best course of action is up to you.
